synapse/lib/lmdbslab.py

```diff
@@ -363,8 +376,97 @@ def _growMapSize(self, size=None):
+        Separate thread loop that manages the prefaulting and locking of the memory backing the data file
+        '''
+        if not s_thishost.get('hasmemlocking'):
+            return
+        MAX_TOTAL_PERCENT = .90  # how much of all the RAM to take
+        MAX_LOCK_AT_ONCE = s_const.gibibyte
+
+        # Calculate a reasonable maximum amount of memory to lock
+
+        s_thisplat.maximizeMaxLockedMemory()
+        locked_ulimit = s_thisplat.getMaxLockedMemory()
+        if locked_ulimit < s_const.gibibyte // 2:
+            logger.warning(
+                'lmdbslab: Operating system limit of maximum amount of locked memory (currently %d) is \n'
+                'too low for optimal performance.', locked_ulimit)
+
+        logger.debug('lmdbslab: memory locking thread started')
+
+        # Note:  available might be larger than max_total in a container
+        max_total = s_thisplat.getTotalMemory()
+        available = s_thisplat.getAvailableMemory()
+
+        PAGESIZE = 4096
+        max_to_lock = (min(locked_ulimit,
+                           int(max_total * MAX_TOTAL_PERCENT),
+                           int(available * MAX_TOTAL_PERCENT)) // PAGESIZE) * PAGESIZE
+
+        path = self.path.absolute() / 'data.mdb'  # Path to the file that gets mapped
+        fh = open(path, 'r+b')
+        fileno = fh.fileno()
+
+        prev_memstart = 0  # The last start of file mapping, used to keep track when the mapping moves
+        prev_memend = 0  # The last end of the file mapping, so we can start from there
+
+        # Avoid spamming messages
+        first_end = True
+        limit_warned = False
+
+        self.resizeevent.set()
+
+        while not self.isfini:
+
+            self.resizeevent.wait()
+            if self.isfini:
+                break
+
+            self.resizeevent.clear()
+
+            memstart, memlen = s_thisplat.getFileMappedRegion(path)
+            if memlen > max_to_lock:
+                memlen = max_to_lock
+                if not limit_warned:
+                    logger.warning('lmdbslab:  memory locking limit reached')
+                    limit_warned = True
+                # Even in the event that we've hit our limit we still have to loop because further mmaps may cause
+                # the base address to change, necessitating relocking what we can
+
+            # The file might be a little bit smaller than the map because rounding (and mmap fails if you give it a
+            # too-long length)
+            filesize = os.fstat(fileno).st_size
+            goal_end = memstart + min(memlen, filesize)
+
+            # If we just started or we evidently got a new base address, restart mapping from the beginning
+            if not prev_memstart or prev_memstart != memstart:
+                prev_memend = prev_memstart = memstart
+
+            assert goal_end >= prev_memend
+
+            # Actually do the prefaulting and locking.  Only do it a chunk at a time to maintain responsiveness.
```

Comment:
what happens if `resizeevent` is set during this loop?

Reply:
It will prefault until it gets to the previous end, whether it was moved or not. If the resizing moved the location, then the memory won't remain locked. If the resizing didn't move the location (which is more likely, the next time around the loop it will continue where it left off.